src/main/java/sparksoniq/spark/iterator/flowr/OrderByClauseSparkIterator.java

```diff
@@ -17,38 +17,151 @@
+        if (_localTupleResults.size() == 0) {
+            this._hasNext = false;
+        } else {
+            this._hasNext = true;
+        }
+    }
+
+
+    /**
+     * Evaluates expressions to atomics(error is thrown if not possible) which are used as keys for sorted TreeMap.
+     * Requires _child iterator to be opened.
+     * @return Sorted TreeMap(ascending). key - atomics from expressions, value - input tuples
+     */
+    private TreeMap<FlworKey,  List<FlworTuple>> mapExpressionsToOrderedPairs() {
+        // tree map keeps the natural item order deduced from an implementation of Comparator
+        // OrderByClauseSortClosure implements a comparator and provides the exact desired behavior for local execution as well
+        TreeMap<FlworKey, List<FlworTuple>> keyValuePairs = new TreeMap<>(new OrderByClauseSortClosure(_expressions, true));
+
+        // assign current context as parent. re-use the same context object for efficiency
+        DynamicContext tupleContext = new DynamicContext(_currentDynamicContext);
+        while (_child.hasNext()) {
+            FlworTuple inputTuple = _child.next();
+
+            List<Item> results = new ArrayList<>(); // results from the expressions will become a key
+            for (OrderByClauseSparkIteratorExpression orderByExpression : _expressions) {
+                tupleContext.removeAllVariables();                     // clear the previous variables
+                tupleContext.setBindingsFromTuple(inputTuple);        // assign new variables from new tuple
+
+                boolean isFieldEmpty = true;
+                RuntimeIterator expression = orderByExpression.getExpression();
+                expression.open(tupleContext);
+                while (expression.hasNext()) {
+                    Item resultItem = expression.next();
+                    if (resultItem != null && !Item.isAtomic(resultItem))
+                        throw new NonAtomicKeyException("Order by keys must be atomics",
+                                orderByExpression.getIteratorMetadata().getExpressionMetadata());
+                    isFieldEmpty = false;
+                    results.add(resultItem);
+                }
+                // if empty ordering field is found, add a Java null as placeholder
+                if (isFieldEmpty) {
+                    results.add(null);
+                }
+                expression.close();
+            }
+            FlworKey key = new FlworKey(results);
+            List<FlworTuple> values = keyValuePairs.get(key);   // all values for a single matching key are held in a list
+            if (values == null) {
+                values = new ArrayList<>();
+                keyValuePairs.put(key, values);
+            }
+            values.add(inputTuple);
+        }
+        return keyValuePairs;
+    }
+
     @Override
     public JavaRDD<FlworTuple> getRDD(DynamicContext context) {
-        if(this._rdd == null) {
+        if (this._child != null) {
```

Comment:
The code here can be simplified (get rid of _rdd, and directly return the final result).